tools/load_system_tablets.hh

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include <filesystem>
+#include <map>
+#include <seastar/core/future.hh>
+
+#include "reader_permit.hh"
+#include "dht/token.hh"
+#include "locator/tablets.hh"
+#include "seastarx.hh"
+
+namespace db {
+class config;
+}
+
+namespace tools {
+
+using tablets_t = std::map<dht::token, locator::tablet_replica_set>;
+
+/// Load the rows of given table in "system.tablets" from its sstables
+///
+/// @param cfg the db config
+/// @param scylla_data_path path to the scylla data directory, which is usually
+///        /var/lib/scylla/data
+/// @param keyspace_name the keyspace name of the table
+/// @param table_name the table name of the table
+/// @param permit the permit for performing read ops
+/// @returns a map from last token to the replica set
+future<tablets_t> load_system_tablets(const db::config& dbcfg,
+                                      std::filesystem::path scylla_data_path,
+                                      std::string_view keyspace_name,
+                                      std::string_view tablet_name,
+                                      reader_permit permit);
```

Comment:
There is already code in `replica/tablets.cc` to parse tablet data into C++ structures. I think it would be better to reuse it. It is not currently exposed (it is hidden in the `replica/tablets.cc`) file, but it should be easy enough to expose a method, which converts `query::result_set_row` to the C++ structures.
Maybe exposing https://github.com/scylladb/scylladb/pull/15541/commits/3d5d3ec80c5afa32cfa8f8e44735ec4e727c9668 is enough.

Reply:
(note, the above commit is from https://github.com/scylladb/scylladb/pull/15541, not merged yet as of the writing of this comment)

Reply:
It was just merged.

Reply:
@denesb i think you are referencing `replica_set_from_row()` instead of `do_load_system_tablets()`. and the closest function in `replica/tablets.cc` is `process_one_row()`, which processes a `cql3::untyped_result_set_row` and updates a `tablet_map`.  `process_one_row()` parses tablets data carried by `cql3::untyped_set_row`, which provides a set of similar but different functions  than `result_set_row` when accessing the cell value by its column name, for instance:

- `result_set_row::get()` 
-  `untyped_result_set_row::get_opt()`

do you suggest adding a type to unify them? i am not sure we have a utility to convert `untyped_result_set_row` to `result_set_row` through. or do you suggest *moving* `replica_set_from_row()` into `replica/tablets.cc` 

another candidate is `deserialize_replica_set()`, which parses a `view_type` after casting it to a `data_value`. we could have a helper which accepts a `data_value` and returns a `tablet_replica_set`. but since you are proposing adding 

>  a method, which converts query::result_set_row to the C++ structures.

i am not quite sure if this is the way to go.

Reply:
I just don't want to duplicate parsing logic of tablet data. We already have all the parsing in `replica/tablets.cc`, I think it would be quite easy to expose a function which can be reused here. I don't insist on how to exactly do this, I'm just trying to reduce the amount of custom code `scylla-sstable` relies on. Best to re-use as much code from scylla core as possible.

Reply:
With all that said, I don't insist at any cost. If you think it is way too much work for what we need, I'm fine with the duplication. Sometimes duplicating some code is the better option.

Reply:
neither do i want to repeat the code. but in this case, i think despite that `replica_set_from_row()` and `do_load_system_tablets()` are repeating what we have in `replica/tablets.cc`, neither of them are reusable enough to be shared at current shape. if we see another use case where we perform the same set of accesses on  `result_set_row` and `untyped_result_set_row`, i will try to come up with a unified solution. 

in this case, i am extracting `tablet_replica_set_from_cell()` to reduce the repeatings.